Add two new optional attributes to the APE 14 WCS: pixel_axis_names and world_axis_names

astrofrog opened the pull request:
Following discussions with @Cadair, this adds two new optional attributes to the APE 14 WCS specification. This is a proposal, and should be approved by the original authors (and interested parties) for the APE 14 document, since the base classes here are now the authoritative reference APIs. So this should not be merged until approved by everyone tagged for review here.

nden commented:
Isn't `world_axis_name` simply `WCS.wcs.cname`?

astrofrog commented:
@nden - yes I think it would map to CNAME for FITS-WCS. But the intent is to have something exposed at the APE14 WCS level that would be the same in e.g. GWCS. Does that make sense?

nden commented:
@astrofrog Yep, I only meant FITS, of course. It makes sense to have it in the APE. I'll review it tomorrow.

nden commented on astropy/wcs/wcsapi/low_level_api.py:
```diff
@@ -330,22 +356,35 @@ def __str__(self):
+            s += (('{0:' + str(world_dim_width) + 'd}').format(iwrl) + '  ' +
+                    ('{0:' + str(world_nam_width) + 's}').format(name) + '  ' +
+                    ('{0:' + str(world_typ_width) + 's}').format(typ) + '  ' +
+                    '{:s}'.format(unit + '\n'))
```
I'd like to propose to NOT define `__str__` at the low level API and leave it to the concrete implementation. I think it would be better to define a new function for this purpose. As I mentioned in #9157, this overwrites the FITS WCS `__str__` and was a bit of a surprise. And it doesn't do anything for `gwcs` as `gwcs` defines its own `__str__`. If it's in a new function which doesn't overwrite something as basic as `__str__`, it will work for both packages using the API and can outline the structure in a common way.
Obviously it's not part of this PR, I mention it because this PR extends `__str__` and all involved are watching the issue.

Cadair replied:
:+1: to implementing `__str__` as a function, I have uses for that as well.

I am in favour of changing the str for `WCS` but it probably needs more thought than how it ended up happening here. This should probably be a separate PR though.

nden commented:
Are these functions going to be added to the official APE before the astropy release?

astrofrog commented:
We stated in the updated APE that the base classes here are now the reference API and I think we won’t update the APE anymore.

nden commented:
@astrofrog Thanks. Good to know.

Cadair commented on astropy/wcs/wcsapi/low_level_api.py:
```diff
@@ -257,16 +258,38 @@ def pixel_bounds(self):
+        """
+        An iterable of strings describing the name for each pixel axis.
+
+        If an axis does not have a name, `None` can be returned.
```
:+1: to empty string here, a single type is generally easier to work with.

astrofrog commented:
I've now implemented ``world_axis_names`` for FITS-WCS using ``CNAME`` and have also updated the API for ``pixel_axis_names`` and ``world_axis_names`` to return strings for axes with no names.

I've moved the str/repr method to a separate function in https://github.com/astropy/astropy/pull/9207 and have rebased this PR on top of https://github.com/astropy/astropy/pull/9207, so https://github.com/astropy/astropy/pull/9207 should be reviewed/merged first then I'll rebase this on master.

eteq commented on astropy/wcs/wcsapi/low_level_api.py:
```diff
@@ -257,16 +258,38 @@ def pixel_bounds(self):
+        """
+        An iterable of strings describing the name for each pixel axis.
+
+        If an axis does not have a name, an empty string should be returned.
```
Maybe this should be reworded to:
```suggestion
        If an axis does not have a name (the default if a subclass does not
        override), an empty string should be returned.
```
Or similar to get at the point that this is *optional*, and if you do nothing everything still works?

eteq commented on astropy/wcs/wcsapi/low_level_api.py:
```diff
@@ -257,16 +258,38 @@ def pixel_bounds(self):
+        """
+        An iterable of strings describing the name for each world axis.
+
+        If an axis does not have a name, an empty string be returned. Note that
```
```suggestion
        If an axis does not have a name (the default if a subclass does not
        override), an empty string should be returned. Note that
```

eteq commented on astropy/wcs/wcsapi/low_level_api.py:
```diff
@@ -257,16 +258,38 @@ def pixel_bounds(self):
+        An iterable of strings describing the name for each world axis.
+
+        If an axis does not have a name, an empty string be returned. Note that
+        these names are just for display purposes and are not standardized.
```
Why does the "just for display" purposes appear here and not in `pixel_axis_names`? Isn't that also true for `pixel_axis_names` just without the added bit about "standardized axis types"?

astrofrog replied:
Done

astrofrog replied:
Done!

bsipocz commented:
Many approvals and passing tests, so I go ahead and merge this.

Thank you @astrofrog!

bsipocz commented:
@astrofrog @pllim - this already seem to have the #9253 issue.
